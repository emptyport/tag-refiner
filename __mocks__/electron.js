export const remote = {
  dialog: {
    showOpenDialog: jest.fn().mockReturnValue('path/to/output folder')
  }
};