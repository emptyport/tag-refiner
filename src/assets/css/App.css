body {
  font-family: Arial, Helvetica, sans-serif;
  font-size: 1.2em;
}

div {
  margin: 5px;
}

h1 {
  font-family: helvetica;
  font-weight: 200;
}

ul {
  margin: 0px;
  padding: 0px;
}

textarea, input {
  border-radius: 10px;
  border: 1px solid black;
  padding: 5px;
}

button {
  border-radius: 10px;
  border: 0px;
  font-size: 1.1em;
  margin: 5px;
  padding: 10px;
  background-color: rgb(0, 72, 180);
  color: white;
}

#start-button {
  background-color: rgb(0, 77, 17);
  color: white;
}

.container {
  border: 5px inset rgb(223, 223, 223);
  border-radius: 10px;
}

.file-entry {
  margin: 10px;
  padding: 0px;
  display: grid;
  grid-template-columns: auto 1fr auto;
  justify-items: center;
  align-items: center;
  grid-column-gap: 10px;
}

.file-progress {
  width: 100%;
  text-align: center;
}

#param-selection {
  display: grid;
  grid-template-columns: 1fr 1fr;
}