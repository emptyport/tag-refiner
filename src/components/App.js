import '../assets/css/App.css';
import React, { Component } from 'react';
const {dialog} = require('electron').remote;
import FileList from './FileList';
let { spectralProcessor } = require('../utils/spectralProcessor');

let presets = require('../assets/presets.json');
console.log("Loaded presets");
console.log(presets);

class App extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      fileList: [],
      msLevel: 3,
      tolerance: 0.002,
      foldChange: 2,
      outputPath: ''
    };

    this.processFiles = this.processFiles.bind(this);
    this.selectFiles = this.selectFiles.bind(this);
    this.selectOutputFolder = this.selectOutputFolder.bind(this);
    this.selectPreset = this.selectPreset.bind(this);
    this.formatListForTextarea = this.formatListForTextarea.bind(this);
    this.changeTolerance = this.changeTolerance.bind(this);
    this.changeMsLevel = this.changeMsLevel.bind(this);
    this.changeFoldChange = this.changeFoldChange.bind(this);
    this.handleResult = this.handleResult.bind(this);
  }

  formatListForTextarea(list) {
    return list.join('\n');
  }

  changeTolerance(event) {
    let tolerance = parseFloat(event.target.value);
    this.setState(function() {
      return {
        tolerance: tolerance
      };
    });
  }

  changeMsLevel(event) {
    let msLevel = parseInt(event.target.value);
    this.setState(function() {
      return {
        msLevel: msLevel
      };
    });
  }

  changeFoldChange(event) {
    let foldChange = parseFloat(event.target.value);
    this.setState(function() {
      return {
        foldChange: foldChange
      };
    });
  }

  handleResult(result, specGen, index) {
    if(result.done) { return; }

    let fileList = this.state.fileList;
    console.log(result.value.progress);
    fileList[index].progress = result.value.progress;
    fileList[index].status = result.value.message;
    this.setState({
      fileList: fileList
    });
    FileList.forceUpdate();
    setTimeout(this.handleResult(specGen.next(), specGen, index), 10);
  }


  processFiles() {
    let rawReporters = document.getElementById("reporter-textarea").value.split("\n");
    let reporters = rawReporters.map(item => parseFloat(item));
    let rawControls = document.getElementById("control-textarea").value.split("\n");
    let controls = rawControls.map(item => parseFloat(item));

    let options = {
      reporters: reporters,
      controls: controls,
      tolerance: this.state.tolerance,
      msLevel: this.state.msLevel,
      foldChange: this.state.foldChange,
      outputPath: this.state.outputPath
    };

    let fileList = this.state.fileList;
    for(let i=0; i<fileList.length; i++) {
    //this.state.fileList.map((file, index) => {
      let file = fileList[i];
      let index = i;
      let specGen = spectralProcessor(file.file, options, index);
      let result = specGen.next();

      /*
      this.handleResult(result, specGen, index);
      let fileList = this.state.fileList;
      fileList[index].progress = 100;
      fileList[index].status = "Done!";
      this.setState({
        fileList: fileList
      });
      
      */

      
      while(!result.done) {
        console.log(result.value.progress);
        fileList[index].progress = result.value.progress;
        fileList[index].status = result.value.message;
        this.setState({
          fileList: fileList
        });
        FileList.forceUpdate();
        result = specGen.next();
      }
      fileList[index].progress = 100;
      fileList[index].status = "Done!";
      

    }
  }

  selectPreset(event) {
    let key = event.target.value;
    console.log(`Preset key selected: ${key}`);

    let reporters, controls;
    if(key==='none') {
      reporters = [];
      controls = [];
    }
    else {
      reporters = presets[key].reporters;
      controls = [presets[key].reporters[0]];
    }

    document.getElementById("reporter-textarea").value = this.formatListForTextarea(reporters);
    document.getElementById("control-textarea").value = this.formatListForTextarea(controls);

  }

  selectOutputFolder() {
    dialog.showOpenDialog({
      properties: [
        'openDirectory'
      ]
    },
    function(folder) {
      console.log(`Output path set to ${folder}`);
      this.setState(function() {
        return {
          outputPath: folder
        };
      });
    }.bind(this));
  }

  selectFiles() {
    dialog.showOpenDialog({ 
      properties: [ 
        'openFile',
        'multiSelections' 
      ],
      filters: [
        { name: 'mzML', extensions: ['mzML'] }
      ]
    }, 
    function(filenames) {
      console.log("Selected files:");
      console.log(filenames);
      let fileList = filenames.map((file, index) => {
        return {
          id: index,
          file: file,
          progress: 0,
          status: "Pending"
        }
      });
      this.setState(function() {
        return{
          fileList: fileList
        }
      });
    }.bind(this));
  }

  render() {
    let presetSelect = [
      {
        name: '---',
        id: 'none'
      }
    ];
    for(let preset in presets) {
      if(presets.hasOwnProperty(preset)) {
        presetSelect.push({
          name: presets[preset].name,
          id: preset
        });
      }
    }

    let presetList = presetSelect.map((preset, index) => {
      return (
        <option key={index} value={preset.id}>{preset.name}</option>
      );
    });


    return (
      <div>

        <div id="file-selection">

          <div className="container">
            <FileList files={this.state.fileList} />
          </div>
          <button onClick={this.selectFiles}>Select files</button>

        </div>

        <div id="param-selection" className="">
          <div id="mass-params-div">
            <div id="preset-div">
              Select a preset: <br />
              <select id="presets" onChange={this.selectPreset}>
                {presetList}
              </select>
            </div>

            <div id="reporter-div">
              <div>Reporter ions:</div>
              <textarea id="reporter-textarea" rows="15" cols="15" defaultValue=''></textarea>
            </div>

            <div id="control-div">
              <div>Control channels:</div>
              <textarea id="control-textarea" rows="5" cols="15" defaultValue=''></textarea>
            </div>
          </div>

          <div id="simple-params-div">
            <div className="param-item">
              <div>Tolerance (Da):</div>
              <input id="tolerance" type="text" value={this.state.tolerance} onChange={this.changeTolerance}/>
            </div>

            <div className="param-item">
              <div>MS Level with reporters</div>
              <input id="msLevel" type="number" min="2" max="3" value={this.state.msLevel} onChange={this.changeMsLevel}/>
            </div>

            <div className="param-item">
              <div>Fold change:</div>
              <input id="foldChange" type="text" value={this.state.foldChange} onChange={this.changeFoldChange}/>
            </div>

            <div className="param-item">
              <div>Output folder:</div>
              <div>{this.state.outputPath}</div>
              <button onClick={this.selectOutputFolder}>Output path</button>
            </div>
          </div>

        </div>

        <button id="start-button" onClick={this.processFiles}>Start</button>
      </div>
    );
  }
}

export default App;
