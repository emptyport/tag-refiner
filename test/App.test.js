import React from 'react';
import App from '../src/components/App';
import renderer from 'react-test-renderer';
import { remote } from 'electron';

test('App initializes correctly', () => {
  const component = renderer.create(
    <App />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  expect(tree.type).toBe('div');
  expect(tree.children.length).toBe(3);
  
  // Checking the fileList div
  let firstChild = tree.children[0];
  expect(firstChild.type).toBe('div');
  expect(firstChild.children.length).toBe(2);
  expect(firstChild.children[0].type).toBe('div');
  expect(firstChild.children[1].type).toBe('button');

  let secondChild = tree.children[1];
  expect(secondChild.type).toBe('div');
  expect(secondChild.children.length).toBe(2);
  expect(secondChild.children[0].type).toBe('div');
  expect(secondChild.children[1].type).toBe('div');

  let thirdChild = tree.children[2];
  expect(thirdChild.type).toBe('button');
  expect(thirdChild.children[0]).toBe('Start');

});

test('App functions', () => {
  const component = renderer.create(
    <App />
  );
  const instance = component.getInstance();
  
  // formatListForTextarea
  let list = ['a', 'b', 'c'];
  let joinedList = list.join('\n');
  let testList = instance.formatListForTextarea(list);
  expect(joinedList).toBe(testList);

  let event;

  // changeTolerance
  event = { target: { value: 0.1 } };
  instance.changeTolerance(event);
  expect(instance.state.tolerance).toBe(0.1);

  // changeMsLevel
  event = { target: { value: 2 } };
  instance.changeMsLevel(event);
  expect(instance.state.msLevel).toBe(2);

  // changeFoldChange
  event = { target: { value: 5 } };
  instance.changeFoldChange(event);
  expect(instance.state.foldChange).toBe(5);

  // open dialog and setting folder and filenames
  instance.selectOutputFolder();
  expect(remote.dialog.showOpenDialog).toHaveBeenCalled();
  instance.selectFiles();
  expect(remote.dialog.showOpenDialog).toHaveBeenCalled();
  instance.setOutputFolder('/some/output/path');
  expect(instance.state.outputPath).toBe('/some/output/path');
  instance.setFileNames(['file1', 'file2']);
  expect(instance.state.fileList).toEqual([
    {
      id: 0,
      file: 'file1',
      progress: 0,
      status: 'Pending'
    },
    {
      id: 1,
      file: 'file2',
      progress: 0,
      status: 'Pending'
    }
  ]);

});