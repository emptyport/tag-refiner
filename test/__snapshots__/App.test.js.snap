// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App initializes correctly 1`] = `
<div>
  <div
    id="file-selection"
  >
    <div
      className="container"
    >
      <div>
        <ul />
      </div>
    </div>
    <button
      onClick={[Function]}
    >
      Select files
    </button>
  </div>
  <div
    className=""
    id="param-selection"
  >
    <div
      id="mass-params-div"
    >
      <div
        id="preset-div"
      >
        Select a preset: 
        <br />
        <select
          id="presets"
          onChange={[Function]}
        >
          <option
            value="none"
          >
            ---
          </option>
          <option
            value="tmt6plex_hcd"
          >
            TMT 6-plex HCD
          </option>
          <option
            value="tmt6plex_etd"
          >
            TMT 6-plex ETD
          </option>
          <option
            value="tmt11plex_hcd"
          >
            TMT 11-plex HCD
          </option>
          <option
            value="tmt11plex_etd"
          >
            TMT 11-plex ETD
          </option>
          <option
            value="itraq4plex_hcd"
          >
            iTraq 4-plex HCD
          </option>
        </select>
      </div>
      <div
        id="reporter-div"
      >
        <div>
          Reporter ions:
        </div>
        <textarea
          cols="15"
          defaultValue=""
          id="reporter-textarea"
          rows="15"
        />
      </div>
      <div
        id="control-div"
      >
        <div>
          Control channels:
        </div>
        <textarea
          cols="15"
          defaultValue=""
          id="control-textarea"
          rows="5"
        />
      </div>
    </div>
    <div
      id="simple-params-div"
    >
      <div
        className="param-item"
      >
        <div>
          Tolerance (Da):
        </div>
        <input
          id="tolerance"
          onChange={[Function]}
          type="text"
          value={0.002}
        />
      </div>
      <div
        className="param-item"
      >
        <div>
          MS Level with reporters
        </div>
        <input
          id="msLevel"
          max="3"
          min="2"
          onChange={[Function]}
          type="number"
          value={3}
        />
      </div>
      <div
        className="param-item"
      >
        <div>
          Fold change:
        </div>
        <input
          id="foldChange"
          onChange={[Function]}
          type="text"
          value={2}
        />
      </div>
      <div
        className="param-item"
      >
        <div>
          Output folder:
        </div>
        <div>
          
        </div>
        <button
          onClick={[Function]}
        >
          Output path
        </button>
      </div>
    </div>
  </div>
  <button
    id="start-button"
    onClick={[Function]}
  >
    Start
  </button>
</div>
`;
